python/ray/air/util/tensor_extensions/arrow.py

```diff
@@ -270,3 +270,282 @@ def to_numpy(self, zero_copy_only: bool = True):
             A single ndarray representing the entire array of tensors.
         """
         return self._to_numpy(zero_copy_only=zero_copy_only)
+
+
+@PublicAPI(stability="beta")
+class ArrowRaggedTensorType(pa.PyExtensionType):
+    """
+    Arrow ExtensionType for an array of heterogeneous-shaped, homogeneous-typed
+    tensors.
+
+    This is the Arrow side of TensorDtype for tensor elements with different shapes.
+    Note that this extension only supports non-ragged tensor elements; i.e., when
+    considering each tensor element in isolation, they must have a well-defined shape.
+
+    See Arrow extension type docs:
+    https://arrow.apache.org/docs/python/extending_types.html#defining-extension-types-user-defined-types
+    """
+
+    def __init__(self, dtype: pa.DataType):
+        """
+        Construct the Arrow extension type for array of heterogeneous-shaped tensors.
+
+        Args:
+            dtype: pyarrow dtype of tensor elements.
+        """
+        super().__init__(
+            pa.struct([("value", pa.list_(dtype)), ("shape", pa.list_(pa.int64()))])
+        )
+
+    def to_pandas_dtype(self):
+        """
+        Convert Arrow extension type to corresponding Pandas dtype.
+
+        Returns:
+            An instance of pd.api.extensions.ExtensionDtype.
+        """
+        from ray.air.util.tensor_extensions.pandas import TensorDtype
+
+        return TensorDtype(
+            self.storage_type["value"].type.value_type.to_pandas_dtype(),
+        )
+
+    def __reduce__(self):
+        return ArrowRaggedTensorType, (self.storage_type["value"].type.value_type,)
+
+    def __arrow_ext_class__(self):
+        """
+        ExtensionArray subclass with custom logic for this array of tensors
+        type.
+
+        Returns:
+            A subclass of pd.api.extensions.ExtensionArray.
+        """
+        return ArrowRaggedTensorArray
+
+    def __str__(self) -> str:
+        dtype = self.storage_type["value"].type.value_type
+        return f"ArrowRaggedTensorType(dtype={dtype})"
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+@PublicAPI(stability="beta")
+class ArrowRaggedTensorArray(pa.ExtensionArray):
```

Comment:
It may be better to call this `ArrowVariableShapeTensorArray` or something to that effect, since a "ragged" tensor supports arbitrary raggedness, e.g. any dimension can be ragged, while this tensor extension has a stricter requirement that all tensor elements must themselves be non-ragged.

Reply:
I didn't call it that since this ragged tensor extension can be ragged in more than a single dimension; e.g., if you have a tensor containing 2 images with shapes (28, 28) and (32, 32), that's ragged in both the height and width dimensions.